packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1379,10 +1379,15 @@ class Navigator extends StatefulWidget {
   /// This callback is responsible for calling [Route.didPop] and returning
   /// whether this pop is successful.
   ///
-  /// The [Navigator] widget should be rebuilt with a [pages] list that does not
-  /// contain the [Page] for the given [Route]. The next time the [pages] list
-  /// is updated, if the [Page] corresponding to this [Route] is still present,
-  /// it will be interpreted as a new route to display.
+  /// This callback is also responsible for updating the list of [Page]s that
+  /// was passed into the [pages]. If the [Page] corresponding to this [Route]
+  /// is still present the next time the [pages] list is updated, it will be
+  /// interpreted as a new route to display.
+  ///
+  /// When updating the list of [Page]s, this callback should avoid triggering
```

Comment:
Not sure what you mean by this. In order to provide a new list of pages you have to rebuild the navigator.

Reply:
In the navigator.pop, it already marks the route entry to be pop, and the route.didpop will trigger the popping animation, at this point the navigator should be happy with its state. Triggering another page rebuild is not necessary, we just need to make sure the 'next' time it updates with the new page list.

The only concern with this is the navigator.pages might still contain the removed page, but it will not be used in any case until a new pages list is provided. If we really want to get rid of it, we might need to change the way we handle popping page route.

Reply:
That rebuild should be cheap, though, since nothing has changed? It is gonna be confusing if the current page list of the navigator doesn't match its internal configuration.

Reply:
I am thinking if we should make developer responsible for rebuilding the navigator with the new pages list, and the navigator just call the onPopPage callback and does not do anything after that?

The document is wrong as well. If user rebuilds the navigator without the removing the page from the pages list, two different outcome can occur.

- If the removed page has not finished popping when the rebuild happens, the page will be updated and continue popping. The result is the page will be popped but still in page list.
- If the removed page has popped when rebuild happens, a new page will be created.

Reply:
You'd have to make sure that the transition delegate actually marks this page for pop then. Seems a little odd.